### PR TITLE
[Mobile App] Error and empty state widgets

### PR DESCRIPTION
**Pull request by NoahNsimbe**
#### Summary of Changes (What does this PR do?)

- [x] Resolves [location errors for android](https://console.firebase.google.com/u/0/project/airqo-250220/crashlytics/app/android:com.airqo.app/issues/f77bb51dda400857058c33afefa48c98?time=last-seven-days&sessionEventKey=6388729100E6000D55CE13481BB2829C_1750555264698637432)
 - [x] Adds error screens / empty pages to 
    - [x] home page
    - [x] analytics
    - [x] KYA
    - [x] Favourite places

#### Status of maturity (all need to be checked before merging):

- [x] I've tested this locally
- [x] I consider this code done
- [ ] This change ready to hit production in its current state
- [ ] The title of the PR states what changed and the related issues number (used for the release note).
- [ ] I've included issue number in the "Closes #ISSUE-NUMBER" part of the "[What are the relevant tickets?](#what-are-the-relevant-tickets)" section to [link the issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).
- [ ] I've updated corresponding documentation for the changes in this PR.
- [ ] I have written unit and/or e2e tests for my change(s).

#### How should this be manually tested?

- [ ] Run the app without internet connection. You should see appropriate error widgets for the home page, analytics, KYA.
- [ ] Delete all favourite places.  You should see an empty state widget on the favourite places page.
- [ ] Modify the API response for air quality readings to an empty list. You should see an empty state widget on the map and  home page. Make sure there are no air quality readings stored locally. You can clear the app's storage or delete the app to achieve this.

#### What are the relevant tickets?

- Closes #<enter issue number here>
- [Jira_card_number]() (If exists)

#### Screenshots (optional)
|  Home page - No internet connection |   Home page - No Air Quality Data (API issues, etc..) |  
|---|---|
| IMG_4109  | IMG_4108  |  
| | |
|  **No Favourite places** |  **No analytics** |  
| IMG_4110  | IMG_4111 |  
| | |
|  **No KYA** | **Requesting location access and showing other African cities**  |  
| IMG_4112  | IMG_4113 |

### CODE REVIEW

File: mobile/lib/blocs/nearby_location/nearby_location_bloc.dart [hunk 1]
```diff
@@ -3,109 +3,62 @@ import 'package:app/services/services.dart';
...
 
-      if (!locationEnabled || !profile.preferences.location) {
-        return;
+      bool? serviceEnabled = await Geolocator.isLocationServiceEnabled();
+      if (!serviceEnabled) {
+        return emit(state.copyWith(
+          blocStatus: NearbyLocationStatus.error,
+          error: NearbyAirQualityError.locationDisabled,
+        ));
       }
```
Comment (NoahNsimbe) [inline mobile/lib/blocs/nearby_location/nearby_location_bloc.dart:43]: Resolving [location errors](https://console.firebase.google.com/u/0/project/airqo-250220/crashlytics/app/android:com.airqo.app/issues/f77bb51dda400857058c33afefa48c98?time=last-seven-days&sessionEventKey=6388729100E6000D55CE13481BB2829C_1750555264698637432)

File: mobile/lib/blocs/dashboard/dashboard_bloc.dart [hunk 2]
```diff
@@ -5,69 +5,109 @@ import 'package:app/services/services.dart';
...
 
 class DashboardBloc extends Bloc<DashboardEvent, DashboardState> {
-  DashboardBloc()
-      : super(const DashboardState(
-          greetings: '',
-          airQualityReadings: [],
-          loading: false,
-        )) {
-    on<UpdateGreetings>(_onUpdateGreetings);
+  DashboardBloc() : super(const DashboardState.initial()) {
```
Comment (NoahNsimbe) [inline mobile/lib/blocs/dashboard/dashboard_bloc.dart:13]: Refactoring home page state management.

File: mobile/lib/constants/config.dart [hunk 3]
```diff
@@ -65,12 +65,10 @@ class Config {
   static String get locationErrorMessage =>
       'Cannot get your location at the moment';
 
-  static int get maxSearchRadius => searchRadius * 2;
-
   static String get playStoreUrl =>
       'https://play.google.com/store/apps/details?id=com.airqo.app';
 
-  static int get searchRadius => 2;
+  static int get searchRadius => 4;
```
Comment (NoahNsimbe) [inline mobile/lib/constants/config.dart:71]: Search radius in kilometers @Baalmart

File: mobile/lib/screens/analytics/analytics_widgets.dart [hunk 4]
```diff
@@ -72,50 +69,6 @@ class AnalyticsAvatar extends StatelessWidget {
   }
 }
 
-class MapAnalyticsMoreInsights extends StatelessWidget {
-  const MapAnalyticsMoreInsights({super.key});
-
```
Comment (NoahNsimbe) [inline mobile/lib/screens/analytics/analytics_widgets.dart:77]: Cleanup. Removing unused classes

File: mobile/lib/screens/dashboard/dashboard_view.dart [hunk 5]
```diff
@@ -51,202 +52,626 @@ class _DashboardViewState extends State<DashboardView> {
...
+    WidgetsBinding.instance.addObserver(this);
+    _listenToStreams();
+  }
+
+  @override
+  void didChangeAppLifecycleState(AppLifecycleState state) {
+    super.didChangeAppLifecycleState(state);
+    switch (state) {
+      case AppLifecycleState.resumed:
+        _refresh();
```
Comment (NoahNsimbe) [inline mobile/lib/screens/dashboard/dashboard_view.dart:321]: Refreshing home page after user has granted the app access to location

File: mobile/lib/screens/dashboard/dashboard_widgets.dart [hunk 6]
```diff
@@ -10,11 +10,92 @@ import 'package:flutter/material.dart';
...
+            height: height,
+          ),
+        );
+      },
+    );
+  }
+}
+
+class DashboardLocationButton extends StatelessWidget {
+  const DashboardLocationButton(this.error, {super.key});
```
Comment (NoahNsimbe) [inline mobile/lib/screens/dashboard/dashboard_widgets.dart:57]: Widget requesting the user to turn on/grant the app location permissions

File: mobile/lib/screens/favourite_places/favourite_places_page.dart [hunk 7]
```diff
@@ -24,30 +24,35 @@ class FavouritePlacesPage extends StatelessWidget {
             if (state.favouritePlaces.isEmpty) {
               context.read<AccountBloc>().add(const RefreshFavouritePlaces());
 
-              return const EmptyFavouritePlaces(); // TODO replace with error page
+              return const NoFavouritePlacesWidget();
```
Comment (NoahNsimbe) [inline mobile/lib/screens/favourite_places/favourite_places_page.dart:27]: Replacing with the new designs for empty states

File: mobile/lib/screens/home_page.dart [hunk 8]
```diff
@@ -219,23 +204,13 @@ class _HomePageState extends State<HomePage> {
   void _onItemTapped(int index) {
     switch (index) {
       case 0:
-        context.read<NearbyLocationBloc>().add(const CheckNearbyLocations());
+        context
+            .read<NearbyLocationBloc>()
+            .add(const SearchLocationAirQuality());
```
Comment (NoahNsimbe) [inline mobile/lib/screens/home_page.dart:209]: Searching for the air quality of current location when a user switches to the home page

File: mobile/lib/screens/profile/profile_widgets.dart [hunk 9]
```diff
@@ -83,7 +83,7 @@ class SignUpSection extends StatelessWidget {
             padding: const EdgeInsets.only(left: 55.0, right: 55.0),
             child: AutoSizeText(
               'Create your account today and enjoy air quality'
-              ' updates and healthtips.',
+              ' updates and health tips.',
```
Comment (NoahNsimbe) [inline mobile/lib/screens/profile/profile_widgets.dart:86]: typo

File: mobile/lib/screens/settings/settings_page.dart [hunk 10]
```diff
@@ -66,6 +66,7 @@ class _SettingsPageState extends State<SettingsPage> {
                         ),
                         child: Column(
                           children: [
+                            // TODO implement app lifecycle
```
Comment (NoahNsimbe) [inline mobile/lib/screens/settings/settings_page.dart:69]: Marking an action item for future implementation.

File: mobile/lib/services/location_service.dart [hunk 11]
```diff
@@ -18,32 +18,27 @@ class LocationService {
...
   }
 
   static Future<Position?> getCurrentPosition() async {
     try {
       return await Geolocator.getCurrentPosition(
-        desiredAccuracy: LocationAccuracy.best,
+        desiredAccuracy: LocationAccuracy.medium,
         forceAndroidLocationManager: true,
-        timeLimit: const Duration(seconds: 10),
+        timeLimit: const Duration(seconds: 20),
```
Comment (NoahNsimbe) [inline mobile/lib/services/location_service.dart:41]: increasing location search timeout.

File: mobile/lib/services/location_service.dart [hunk 12]
```diff
@@ -187,6 +158,18 @@ class LocationService {
...
+        await PermissionService.checkPermission(AppPermission.location);
+
+    if (!permissionGranted) {
+      return;
+    }
+
+    bool? serviceEnabled = await Geolocator.isLocationServiceEnabled();
+    if (!serviceEnabled) {
+      return;
+    }
```
Comment (NoahNsimbe) [inline mobile/lib/services/location_service.dart:171]: Checking if a user has permitted the app to access current location and has turned on location

File: mobile/lib/utils/extensions.dart [hunk 13]
```diff
@@ -171,18 +171,40 @@ extension AirQualityReadingListExt on List<AirQualityReading> {
...
+
+    final List<String> countries = data.map((e) => e.country).toSet().toList();
+    countries.shuffle();
+    while (data.isNotEmpty) {
+      for (final country in countries) {
+        List<AirQualityReading> countryReadings = data
+            .where((element) => element.country.equalsIgnoreCase(country))
+            .take(1)
+            .toList();
+        shuffledData.addAll(countryReadings);
```
Comment (NoahNsimbe) [inline mobile/lib/utils/extensions.dart:189]: Adding other African countries to home page

File: mobile/lib/widgets/error_widgets.dart [hunk 14]
```diff
@@ -98,6 +99,179 @@ class NoAirQualityDataWidget extends StatelessWidget {
...
+          const Spacer(),
+        ],
+      ),
+    );
+  }
+}
+
+class NoKyaWidget extends StatelessWidget {
+  const NoKyaWidget({super.key, required this.callBack});
+  final Function() callBack;
```
Comment (NoahNsimbe) [inline mobile/lib/widgets/error_widgets.dart:225]: Empty KYA widget

File: mobile/lib/widgets/error_widgets.dart [hunk 15]
```diff
@@ -98,6 +99,179 @@ class NoAirQualityDataWidget extends StatelessWidget {
...
+          ),
+          const Spacer(),
+        ],
+      ),
+    );
+  }
+}
+
+class NoAnalyticsWidget extends StatelessWidget {
+  const NoAnalyticsWidget({super.key, required this.callBack});
```
Comment (NoahNsimbe) [inline mobile/lib/widgets/error_widgets.dart:175]: Empty analytics widget

File: mobile/lib/widgets/error_widgets.dart [hunk 16]
```diff
@@ -98,6 +99,179 @@ class NoAirQualityDataWidget extends StatelessWidget {
   }
 }
 
+class NoFavouritePlacesWidget extends StatelessWidget {
+  const NoFavouritePlacesWidget({super.key});
+
```
Comment (NoahNsimbe) [inline mobile/lib/widgets/error_widgets.dart:104]: Empty Favourite places widget

File: mobile/lib/widgets/custom_shimmer.dart [hunk 17]
```diff
@@ -16,6 +18,17 @@ class CircularLoadingIndicator extends StatelessWidget {
...
+    if (Platform.isAndroid) {
+      return SizedBox(
+        height: 8,
+        width: 8,
+        child: CircularProgressIndicator(
+          strokeWidth: 1,
+          color: CustomColors.appColorBlue,
+        ),
+      );
+    }
```
Comment (NoahNsimbe) [inline mobile/lib/widgets/custom_shimmer.dart:30]: android loading indicator

File: mobile/lib/constants/config.dart (hunk 3 above)
Comment (Baalmart) [reply mobile/lib/constants/config.dart:72]: Thanks @NoahNsimbe . 4 KM is good. 
If something is beyond 4 KM, then it is definitely not nearby.

**Review by Baalmart — APPROVED** [review]
This LGTM, thanks @NoahNsimbe

**Review by BenjaminSsempala — COMMENTED** [review]
LGTM

**Pull request merged**